registry/storage/paths.go

```diff
@@ -243,6 +248,20 @@ func pathFor(spec pathSpec) (string, error) {
 		return path.Join(append(repoPrefix, v.name, "_uploads", v.id, "hashstates", string(v.alg), offset)...), nil
 	case repositoriesRootPathSpec:
 		return path.Join(repoPrefix...), nil
+	case subjectReferrerLinkPathSpec:
+		manifestPath, err := pathFor(manifestRevisionPathSpec{name: v.name, revision: v.subject})
+		if err != nil {
+			return "", err
+		}
+
+		artifactType := "_" + url.QueryEscape(v.mediaType)
```

Comment:
How about doing an md5 sum here to prevent artifactType from containing special characters or artifactType being too long?

Reply:
`artifactType := fmt.Sprintf("%x", md5.Sum([]byte(v.mediaType)))`

Reply:
A hash is not a bad idea. @Jamstah and I did [discuss](https://github.com/distribution/distribution/issues/3716#issuecomment-1382083234) this and decided to use underscore prefixed URL encoded artifactType.

When listing referrers, we have to read the referrer manifests anyway to get the annotations, so a hash is no less efficient in that respect.

MD5 would be a weird choice, it would light up linters as an insecure hash. Would a collision be a security issue? Listing referrers by artifactType could yield an artifact of a different type, might have downstream impact on clients? We could use sha256 but we'd probably have to do the `<algorithm>/<digest>` directory structure to support more algorithms when sha256 falls, and then we'd have to enumerate another directory when listing.

It would remove directory name length concerns, for the storage back ends which actually have directories and have limited name lengths.

It would obfuscate the artifactType from admins browsing the back end storage, that's quite a minimal concern.

I'm open to the idea if done right.